secp384r1 Solinas' Implementation and Power asm acceleration

rohanmclure opened the pull request:
Adopt a 56-bit redundant-limb implementation, utilising Solinas' reduction for the NIST p384 curve. In keeping with the ecp_nistp224.c and ecp_nistp521.c implementations, allow for future processor-specific acceleration of  felem_{square,mul} methods by means of a simple ABI.

This implementation yielded performance gains of 446% and 106% on digital signing and verification respectively (tested on the Apple M1 Max), and substantial gains on all other platforms tested.

Also provide an asm implementation for Power processors P9 and above, yielding an additional speedup of 9.5% on signature generation.

rohanmclure commented:
This PR updates the documentation on `EC_GFp_nistp384_method()`, as these functions should all be deprecated. Doing so would appear to be to introduce an anachronism. If anyone has any suggestions on how to avoid this I'd appreciate them.

t8m commented on include/openssl/ec.h:
```diff
@@ -144,6 +144,11 @@ OSSL_DEPRECATEDIN_3_0 const EC_METHOD *EC_GFp_nistp224_method(void);
  */
 OSSL_DEPRECATEDIN_3_0 const EC_METHOD *EC_GFp_nistp256_method(void);
 
+/** Returns 64-bit optimized methods for nistp384
+ *  \return  EC_METHOD object
+ */
+OSSL_DEPRECATEDIN_3_0 const EC_METHOD *EC_GFp_nistp384_method(void);
+
```
Just avoid adding the public function and create an internal non-deprecated function. IMO there is no need to have this function public.

rohanmclure replied:
Done! Thanks

t8m commented on crypto/ec/ec_local.h:
```diff
@@ -551,6 +555,27 @@ int ossl_ec_GFp_nistp256_points_mul(const EC_GROUP *group, EC_POINT *r,
+                                    const BIGNUM *scalars[], BN_CTX *ctx);
+int ossl_ec_GFp_nistp384_precompute_mult(EC_GROUP *group, BN_CTX *ctx);
+int ossl_ec_GFp_nistp384_have_precompute_mult(const EC_GROUP *group);
+const EC_METHOD *EC_GFp_nistp384_method(void);
```
All the newly added internal non-static functions should be with lower case ossl_<module> prefix - i.e. `ossl_ec_GFp_nistp384_method` and similarly also for those two other functions added above.

If you want, you can also rename the other internal EC_ functions in a separate commit, to make the function name style uniform. But that is not required for this PR.

beldmit commented on crypto/ec/ec_lib.c:
```diff
@@ -99,12 +99,16 @@ void EC_pre_comp_free(EC_GROUP *group)
 #else
     case PCT_nistp224:
     case PCT_nistp256:
+	case PCT_nistp384:
```
Nit: alignment

beldmit commented on crypto/ec/ec_local.h:
```diff
@@ -264,12 +265,13 @@ struct ec_group_st {
     union {
         NISTP224_PRE_COMP *nistp224;
         NISTP256_PRE_COMP *nistp256;
+		NISTP384_PRE_COMP *nistp384;
```
Nit: alignment

rohanmclure replied:
Corrected, thanks.

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+ * ECDSA low level APIs are deprecated for public use, but still ok for
+ * internal use.
+ */
+#include "internal/deprecated.h"
```
Why?  This is a new file, it should even mention deprecation.

t8m replied:
Well it could use some api internally that is deprecated, however I do not see any such api in use here.

rohanmclure replied:
Apologies, this is residual from v1, where I had deprecation annotations on ossl_ec_nistp384_method().

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+
+static void felem_one(felem out)
+{
+    unsigned i;
```
Our coding style uses `unsigned int` here and elsewhere.

Also a blank line separating declarations and code, again here and elsewhere.

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+    unsigned i;
+    out[0] = 1;
+    for (i = 1; i < NLIMBS; i++)
+        out[i] = 0;
```
Why not use `memset(3)` ?

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+{
+    unsigned i;
+    for (i = 0; i < NLIMBS; i++)
+        out[i] = in[i];
```
Why not use `memcpy(3)` ?

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+ */
+static void felem_neg(felem out, const felem in)
+{
+    /* In order to prevent underflow, we subtract from 0 mod p. */
```
More explanation would be nice.  The subtraction doesn't look to be from zero.
_subtract from p which is 0 mod p_ perhaps?

rohanmclure replied:
I meant subtract from the congruence class containing zero. Let's change that to 'subtract from a multiple of p'.

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+static void felem_diff64(felem out, const felem in)
+{
+    /*
+     * In order to prevent underflow, we add 0 mod p before subtracting.
```
Again, the comment could be improved.

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+static ossl_inline void felem_square_reduce(felem out, const felem in)
+{
+    widefelem tmp;
+    felem_square(tmp, in);
```
blank line above -- blank line to separate declarations and code.

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+    felem_mul_reduce(ftmp, in, ftmp);    /* 2^383 + ... + 2^129 + 2^127 + ... + 2^96 + 2^31 + ... + 2^2 + 2^0 */
+
+    for (i = 0; i < NLIMBS; i++)
+        out[i] = ftmp[i];
```
`memcpy(3)`?

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+static limb felem_is_zero(const felem in)
+{
+    limb zero, p384;
+    zero = in[0] | in[1] | in[2] | in[3] | in[4] | in[5] | in[6];
```
blank line above.

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+{
+    static const int64_t two56 = ((limb) 1) << 56;
+
+    /* We know for a fact that 0 <= in < 2*p, for p = 2^384 - 2^128 - 2^96 + 2^32 - 1
```
Comment style nit:
```c
/*
 * We know for a fact that 0 <= in < 2*p, for p = 2^384 - 2^128 - 2^96 + 2^32 - 1
 * Perform two successive, idempotent subtractions to reduce if in >= p.
 */
```

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+    tmp[5] &= 0x00ffffffffffffff;
+
+    for (i = 0; i < NLIMBS; i++)
+       out[i] = tmp[i];
```
`memcpy(3)`?

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+ *
+ * Building on top of the field operations we have the operations on the
+ * elliptic curve group itself. Points on the curve are represented in Jacobian
+ * coordinates */
```
style nit: close the comment on a new line.

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+ *   http://hyperelliptic.org/EFD/g1p/auto-shortw-jacobian-3.html#doubling-dbl-2001-b
+ *
+ * Outputs can equal corresponding inputs, i.e., x_out == x_in is allowed.
+ * while x_out == y_in is not (maybe this works, but it's not tested). */
```
style nit: close the comment on a new line.

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+    for (i = 0; i < NLIMBS; i++) {
+        const limb tmp = mask & (in[i] ^ out[i]);
+        out[i] ^= tmp;
+    }
```
Is it worthwhile using our constant time #include for this?
Not a firm request but worth thinking about.

rohanmclure replied:
There aren't any real fit-for-purpose methods there. For example to use `constant_time_cond_swap_buff` we would need to copy the contents of `in` first. This method occurs in a hot path, so keeping original implementation.

paulidale commented on crypto/ec/ecp_nistp384.c:
nit: close comment on new line.

paulidale replied:
This one was skipped.

paulidale commented on crypto/ec/ecp_nistp384.c:
blank line separating code from declarations.

paulidale commented on crypto/ec/ecp_nistp384.c:
Again, would our constant time support be useful here?

rohanmclure replied:
As `select_point` is used in a hot path, and so halves the performance of generating digital signatures unfortunately. Keeping original implementation.

paulidale commented on crypto/ec/ecp_nistp384.c:
Indentation looks off here.

paulidale commented on crypto/ec/ecp_nistp384.c:
indentation??

paulidale commented on crypto/ec/ecp_nistp384.c:
indentation

paulidale commented on crypto/ec/ecp_nistp384.c:
Need to free `p->references` using `CRYPTO_FREE_REF`.
On most platforms this is a no-op but not on all.

paulidale commented on crypto/ec/ecp_nistp384.c:
blank line above.

paulidale commented on crypto/ec/ecp_nistp384.c:
We'd not normally have the inner parenthesis, here and elsewhere.

rohanmclure commented:
Thanks @t8m , @beldmit , @paulidale for your comments thus far. I will get to them at the beginning of next week, but am on leave until then.

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1937 @@
+    output[6] += output[5] >> 56;   /* output[6] < 2^48 + 1 */
+    output[5] &= 0x00ffffffffffffff;
+
+    for (i = 0; i < NLIMBS; i++)
```
`memcpy(3)` ?

rohanmclure replied:
Members in `output` and `out` have different sizes, so I'll keep the loop of assignments here.

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1937 @@
+    int64_t tmp[NLIMBS], a;
+    unsigned int i;
+
+    for (i = 0; i < NLIMBS; i++)
```
`memcpyt(3)` ?

paulidale commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1937 @@
+     * Case 2: a = all ones if p <= in < 2^384, 0 otherwise
+     */
+
+    int64_t cond[5];
```
Please move this up to the beginning of the function, we don't mix code and declarations.

slontis commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1935 @@
+ * out[i] < 2^56 for i < 6,
+ * out[6] <= 2^48
+ */
+static void felem_reduce(felem out, const widefelem in)
```
Is there any chance of explaining how you come up with all of the numbers?

rohanmclure replied:
Are we talking about the inequalities or about the program constants?

When it comes to the subtraction constants, they are chosen as power of two multiples of $p_{384}$ that are sufficiently large so that each limb can be sufficiently large. This is achieved through a telescopic sum. You'll notice that we subtract $2^{68}$ from a high limb whenever we add $2^{124}$ to a lower limb. Since $2^{124} = 2^{56} \cdot 2^{68}$ these two alterations cancel out, but lead to each limb being roughly $2^{124}$ in size. This is helpful, as adding a multiple of our modulus like so allows us to avoid underflow when subtracting limbs, which is a requirement for correctness.

rohanmclure commented:
Just updated `felem_reduce` to make `output[6]` larger and fix my bounds checks.

slontis commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1931 @@
+/*
+ * The underlying field. P384 operates over GF(2^384-2^128-2^96+2^32-1). We
+ * can serialize an element of this field into 48 bytes where the most
+ * significant byte contains only a single bit. We call this an felem_bytearray.
```
Is the 'Most significant byte contains only a single bit' correct? (That would be 377 bits?)

rohanmclure replied:
Apologies, yes this comment is completely wrong. I got this confused with `felem_reduce`, which returns a 385 bit number.

slontis commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1931 @@
+static void felem_neg(felem out, const felem in)
+{
+    /* In order to prevent underflow, we subtract from a multiple of p. */
+    static const limb two60m52m4 = (((limb) 1) << 60)
```
I am still struggling to get my head around how you come up with these static constants.. Can you give some insight into how you come up with the numbers below.. What is the actual multiple of p here? (x % p = 0 should be zero)..
I can see that the range seems top be 56 bits with the 2^60-2^4 part (with zeros in the top and bottom bits).

rohanmclure replied:
In `felem_neg` it's just $2^{12} p_{384}$. See $2^{384} \cdot 2^{12} = 2^{60} \cdot (2^{56})^6$. We then use telescopic sums to introduce a $2^{60}$ term in every limb.

In each routine I use a simple power of two to get our multiple. Hope that makes sense.

rohanmclure replied:
I may as well do this one in full:
$2^{12} p_{384} = 2^{396} - 2^{140} - 2^{108} + 2^{44} - 2^{12} = 2^{60} (2^{56})^6 - 2^{28} (2^{56})^2 - 2^{52} \cdot 2^{56} + 2^{44} - 2^{12}$
Now introducing into limbs 0 .. 5 a $2^60$ term:
$= (2^{60} - 2^4) (2^{56})^6 + (2^{60} - 2^4) (2^{56})^5 + (2^{60} - 2^4) (2^{56})^4 + (2^{60} - 2^4) (2^{56})^3 + (2^{60} - 2^{28} - 2^4) (2^{56})^2 + (2^{60} - 2^{52} - 2^4) 2^{56} + (2^{60} + 2^{44} - 2^{12})$

slontis replied:
Thanks for explaining (and your patience).. That makes sense..
Is felem_diff_128_64 similiar?
Would you maybe be able to add comments explaining this in multiple functions, so others can also understand what is going on.

rohanmclure replied:
Not a problem. Yep I was basically just copying the style of ecp_nistp{224,521}.c, but it doesn't hurt to just state the multiple up front.

Precisely, in `felem_diff_128_64` it's $2^{16}p_{384}$, and in `felem_reduce` it's $2^{76}p_{384}$.

slontis replied:
Is the felem_diff_128_64 still 2^56 based

rohanmclure replied:
Yes, so irrespective of the machine width - `widefelem` or `felem` and irrespective of the number ($k$) of limbs, the value being represented is always:

$\sigma^{k}_{i=0} (2^{56})^i l_i$

slontis commented:
I brought this PR up in the OTC meeting this week.
Hopefully @romen can also have a look at it. Most of it is similar to the other variants.
It would be easier to review if you were to explain in comments, how you came up with the values for the static constants, as you have done in the review conversation.
I wonder if there is some way to thrash test this, to make sure we don't have any 'bounds' problems. @romen mentioned there may have been issues previously with the other implementations.

slontis commented:
Thank you for doing this PR. I wanted to add it, but didnt understand the maths in the other implementations..

I am wondering if others have opinions about if there should be multiple config options,
instead of just enable-ec_nistp_64_gcc_128?

Ideally p384 and p521 could be on by default (on systems that have support) since they are much faster...
The p256 case seems to be slightly slower.

(This doesnt have to be part of this PR)

rohanmclure commented:
Yep good suggestion, I'll include these as comments.

I've written a suite of random tests with Python's `cffi` if that would help. Even still, it's best combined with the in-tree and external test suites, which include some important corner cases. Could something like this be useful?

rohanmclure commented:
> Yep good suggestion, I'll include these as comments.
> 
> I've written a suite of random tests with Python's `cffi` if that would help. Even still, it's best combined with the in-tree and external test suites, which include some important corner cases. Could something like this be useful?



> I brought this PR up in the OTC meeting this week.
> Hopefully @romen can also have a look at it. Most of it is similar to the other variants.
> It would be easier to review if you were to explain in comments, how you came up with the values for the static constants, as you have done in the review conversation.

Yep good suggestion, I'll include these as comments.

> I wonder if there is some way to thrash test this, to make sure we don't have any 'bounds' problems. @romen mentioned there may have been issues previously with the other implementations.

I've written a suite of random tests with Python's cffi if that would help. Even still, it's best combined with the in-tree and external test suites, which include some important corner cases. Could something like this be useful?

rohanmclure commented:
Updated license, unrolled loop for clarity in `felem_diff128`, and commented underflow constants.

slontis commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1975 @@
+static BIGNUM *felem_to_BN(BIGNUM *out, const felem in)
+{
+    felem_bytearray b_out;
+    felem_to_bin48(b_out, in);
```
insert a blankline after the variable declarations

slontis commented:
For anyone else trying to follow this the telescopic sum is given by:
$$\sum_{i=1}^n \left( 2^a - 2^b \right) \left( 2^\left( a-b \right)) \right)^i  = 2^\left(a+n(a-b)\right) - 2^a$$

slontis commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1975 @@
+    output[5] += two124m68;
+    output[6] += two124m68;
+
+    /* eliminate in[6], ..., in[12] */
```
Can you expand this comment to explain what the code below is doing.. I get the carry part but not this part..

rohanmclure replied:
I'm deciding the most succinct way to comment this in code, but the underlying principle is this.
Let $p_{384} = 2^{384} - \delta$, then it follows that $2^{384} \equiv \delta \mod{p_{384}}$.

Thus for any suitable choice of $X$ and $Y$ such that our input is equal (or congruent) to $X + 2^{384}Y$ [1] which is in turn congruent to $X + \delta Y$.

The way you start off eliminating the high limbs (it should read 'eliminate in[7], ...., in[12]` is precisely with this substitution.

```c
    /* X += 2^128 Y */
    output[8] += in[12] >> 32;
    output[7] += (in[12] & 0xffffffff) << 24;

    /* X += 2^96 Y */
    output[7] += in[12] >> 8;
    output[6] += (in[12] & 0xff) << 48;
    
    /* X += (-2^32 + 1) Y */
    output[6] -= in[12] >> 16;
    output[5] -= ((in[12] & 0xffff) << 40);
    output[6] += in[12] >> 48;
    output[5] += (in[12] & 0xffffffffffff) << 8;
```

The reason why at each limb, and for term of $\delta = 2^{128} + 2^{96} + (-2^{32} + 1)$ there are multiple operations being performed is simply to reduce the maximum size of operands to prevent under/overflow.

in reality we do at least three such substitutions in this code. Firstly we use limbs `in[12]`, `in[11]`, `in[10]`, `in[9]` to be our $Y$. Then afterwards we take `output[8]`, `output[7]` as our $Y$ (the result of the previous substitution accumulated into these), eliminating these two limbs. Finally, we let `temp` be our choice of $Y$ (each of these with suitable scaling so that [1] holds) and substitute it (after partially carrying into `temp` in order to minimise the size of the resulting, reduced number). We perform one last final carry and then obtain the desired bound.

slontis commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1975 @@
+ * https://www.openssl.org/source/license.html
+ */
+
+/* Copyright 2023 IBM Corp.
```
/*
 * Copyright

paulidale replied:
I don't see why not.  There are plenty of other files with dual copyrights.

slontis commented on crypto/ec/ecp_nistp384.c:
```diff
@@ -0,0 +1,1975 @@
+       cond[i] = ((cond[i] & 0x00ffffffffffffff) - 1) >> 63;
+
+    /*
+     * The condition for determining whether in is greather than our
```
greater

slontis commented:
Apart from some minor NITS this looks good to me.

rohanmclure commented:
Thanks for your comments. I've updated `felem_reduce`, giving it a header-comment providing an overview of the elimination stages, extended the in-line comments and replaced the `output` variable with `acc`.

paulidale reviewed (approved):
Nice work and thanks for persevering.

openssl-machine commented:
This pull request is ready to merge

tmshort commented:
I'd merge this, but @beldmit has an outstanding "change requested".

tmshort commented:
Merged to `master`, thank you for your contribution!
